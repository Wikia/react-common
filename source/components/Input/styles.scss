@import "~@fandom/design-system/dist/scss/wds-variables/index.scss";
@import "~@fandom/design-system/dist/scss/wds-components/inputs/_input.scss";
