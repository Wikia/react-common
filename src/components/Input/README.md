An empty input:
```js
<Input label="An empty input" />
```

An empty email input with hint:
```js
<Input label="E-mail" type="email" hint="Type your e-mail" />
```
Other available types are: `text` (default), `number`, `email`, `search`, `tel`, `url` and `password`.

A non-empty input:
```js
<Input label="A non-empty input" value="lorem ipsum" />
```

Disabled input:
```js
<Input label="Disabled input" value="lorem ipsum" disabled />
```

Readonly input:
```js
<Input label="Readonly input" value="lorem ipsum" readonly />
```

A non-empty input with the error state:
```js
<Input label="Some input" value="lorem ipsum" status="error" hint="Need something else" />
```

An non-empty textarea:
```js
<Input label="Multiline input" value="This is a value" type="multiline" rows="3" />
```

An resizeable textarea:
```js
<Input label="Multiline resizeable textarea" value="This is a value" type="multiline" rows="2" resize />
```

An autoresizeable textarea:
```js
<Input label="Multiline autoresizeable textarea" value="This is a value" type="multiline" resize="auto" />
```
