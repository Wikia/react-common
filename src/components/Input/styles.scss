@import "~design-system/dist/scss/wds-variables/index.scss";
@import "~design-system/dist/scss/wds-components/inputs/_input.scss";
