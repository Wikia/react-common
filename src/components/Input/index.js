import React from 'react';
import PropTypes from 'prop-types';
import uniqueId from 'lodash.uniqueid';

import './styles.scss';

function generateId() {
  return uniqueId('wds_input_');
}

class Input extends React.Component {
  constructor(props) {
    super(props);

    const value = props.value;
    const id = props.id || generateId();

    this.state = {
      value,
      id,
      isEmpty: value.length === 0,
      isFocused: false,
      dynamicTextareaHeight: null,
    };

    this.handleBlur = this.handleBlur.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleFocus = this.handleFocus.bind(this);
    this.handleAutoResize = this.handleAutoResize.bind(this);
  }

  componentDidMount() {
    this.autoFocus();
  }

  componentWillReceiveProps(newProps) {
    const id = newProps.id || generateId();

    this.setState({
      value: newProps.value,
      id,
      isEmpty: newProps.value.length === 0,
    });
  }

  componentDidUpdate() {
    this.forceFocus();
  }

  getClassName() {
    let statusClass = null;

    if (this.state.isFocused) {
      statusClass = 'is-focused';
    } else if (this.state.isEmpty) {
      statusClass = 'is-empty';
    }

    return [
      'wds-input',
      this.props.disabled ? 'is-disabled' : null,
      this.props.readonly ? 'is-readonly' : null,
      this.props.status === 'error' ? 'has-error' : null,
      typeof this.props.resize === 'boolean' && this.props.resize ? 'is-resize' : null,
      statusClass,
      this.props.className,
    ].filter(Boolean).join(' ');
  }

  getInputClassName() {
    return [
      'wds-input__field',
      this.props.inputClassName,
    ].join(' ');
  }

  getLabelClassName() {
    return [
      'wds-input__label',
      this.props.labelClassName,
    ].join(' ');
  }

  getHintClassName() {
    return [
      'wds-input__hint',
      this.props.hintClassName,
    ].join(' ');
  }

  getSharedInputProps() {
    return {
      className: this.getInputClassName(),
      id: this.state.id,
      name: this.state.id,
      value: this.state.value,
      onChange: this.handleChange,
      onBlur: this.handleBlur,
      onFocus: this.handleFocus,
      onKeyUp: this.props.onKeyUp,
      onKeyDown: this.props.onKeyDown,
      onKeyPress: this.props.onKeyPress,
      onPaste: this.props.onPaste,
      readOnly: this.props.readonly,
      disabled: this.props.disabled,
      tabIndex: this.props.tabIndex,
      placeholder: this.props.placeholder,
    };
  }

  focus() {
    if (this.input && document.activeElement !== this.input) {
      this.input.focus();
    }
  }

  blur() {
    if (this.input) {
      this.input.blur();
    }
  }

  handleChange(event) {
    if (this.props.readonly || this.props.disabled) {
      return;
    }

    const value = event.target.value;

    this.setState({
      value,
      isEmpty: value.length === 0,
    });
    this.props.onChange(value, event);
  }

  handleFocus(event) {
    if (this.props.readonly) {
      return;
    }

    this.setState({
      isFocused: true,
    });
    this.props.onFocus(event);
  }

  handleBlur(event) {
    this.setState({
      isFocused: false,
    });
    this.props.onBlur(event);
  }

  isAutoFocus() {
    return (this.props.autoFocus || this.props.forceFocus) && !this.props.disabled && !this.props.readonly;
  }

  isForceFocus() {
    return this.props.forceFocus && !this.props.disabled && !this.props.readonly;
  }

  isAutoResize() {
    return this.props.resize === 'auto' && !this.props.disabled && !this.props.readonly;
  }

  autoFocus() {
    if (this.isAutoFocus() && this.input) {
      this.input.focus();
    }
  }

  forceFocus() {
    if (this.isForceFocus() && this.input) {
      if (document.activeElement !== this.input) {
        this.input.focus();
      }
    }
  }

  handleAutoResize() {
    // height has to be reset first because if not it keeps increasing every time user will type a character
    // setting actual height must be done in setState callback, because React might optimize this into one setState call
    // scrollHeight includes padding but not border, we need to compensate this to avoid slight height change
    // keep value in sync with bottom-border in .wds-input__field styles
    const BOTTOM_BORDER_WIDTH = 1;

    this.setState({dynamicTextareaHeight: 'auto'}, () => {
      this.setState({dynamicTextareaHeight: `${this.input.scrollHeight + BOTTOM_BORDER_WIDTH}px`});
    });

    // to prevent scroll jumping
    this.input.scrollTop = this.input.scrollHeight;
  }

  renderMultiline() {
    const props = {
      ...this.getSharedInputProps(),
      rows: this.props.rows,
    };

    if (this.isAutoResize()) {
      props.onInput = this.handleAutoResize;
    }

    if (this.state.dynamicTextareaHeight) {
      props.style = {height: this.state.dynamicTextareaHeight};
    }

    return (
      <textarea ref={(input) => { this.input = input; }} {...props}>
        {this.state.value}
      </textarea>
    );
  }

  renderInput() {
    if (this.props.type === 'multiline') {
      return this.renderMultiline();
    }

    const props = {
      ...this.getSharedInputProps(),
      type: this.props.type,
    };

    return <input ref={(input) => { this.input = input; }} {...props} />;
  }

  renderLabel() {
    return (
      <label className={this.getLabelClassName()} htmlFor={this.state.id}>
        {this.props.label}
      </label>
    );
  }

  renderHint() {
    if (!this.props.hint) {
      return null;
    }

    return <div className={this.getHintClassName()}>{this.props.hint}</div>;
  }

  render() {
    return (
      <div className={this.getClassName()}>
        {this.renderInput()}
        {this.renderHint()}
        {this.renderLabel()}
      </div>
    );
  }
}

Input.propTypes = {
  /**
   * Additional class name for the component
   */
  className: PropTypes.string,
  /**
   * Additional class name for the hint
   */
  hintClassName: PropTypes.string,
  /**
   * Additional class name for the input
   */
  inputClassName: PropTypes.string,
  /**
   * Additional class name for the label
   */
  labelClassName: PropTypes.string,
  /**
   * ID of the element - by default it's generated automatically
   */
  id: PropTypes.string,
  /**
   * Type of the input.
   * Use `multiline` for multi-line input (textarea).
   */
  type: PropTypes.oneOf(['text', 'number', 'email', 'search', 'tel', 'url', 'password', 'multiline']),
  /**
   * Value
   */
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  /**
   * Label that we want to display.
   */
  label: (props, propName) => {
    if (props.placeholder && props[propName]) {
      return new Error(`Prop ${propName} is not used when placeholder is set`);
    }

    if (!props.placeholder && !props[propName]) {
      return new Error(`Prop ${propName} is required when placeholder is not set`);
    }

    if (typeof props[propName] !== 'string') {
      return new Error(`Prop ${propName} is not a string`);
    }

    return null;
  },
  /**
   * Hint to display
   */
  hint: PropTypes.string,
  /**
   * Placeholder to display
   */
  placeholder: PropTypes.string,
  /**
   * Status
   */
  status: PropTypes.oneOf(['normal', 'error']),
  /**
   * Tab Index
   */
  tabIndex: PropTypes.number,
  /**
   * Initial number of rows
   *
   * **Note**: This prop only makes sense for multiline inputs.
   */
  rows: PropTypes.number,
  /**
   * Can the textarea be resized by the user
   * Use `auto` to adjust textarea height automatically
   *
   * **Note**: This prop only makes sense for multiline inputs.
   */
  resize: PropTypes.oneOf(['auto', true, false]),
  /**
   * Auto focus flag
   */
  autoFocus: PropTypes.bool,
  /**
   * Force focus flag
   */
  forceFocus: PropTypes.bool,
  /**
   * Disabled flag
   */
  disabled: PropTypes.bool,
  /**
   * Readonly flag
   */
  readonly: PropTypes.bool,
  /**
   * Callback for `onBlur` event
   */
  onBlur: PropTypes.func,
  /**
   * Callback for `onChange` event - will be called whenever the value chnages
   * with `callback(value, event)`.
   */
  onChange: PropTypes.func,
  /**
   * Callback for `onFocus` event
   */
  onFocus: PropTypes.func,
  /**
   * Callback for `onKeyDown` event
   */
  onKeyDown: PropTypes.func,
  /**
   * Callback for `onKeyPress` event
   */
  onKeyPress: PropTypes.func,
  /**
   * Callback for `onKeyUp` event
   */
  onKeyUp: PropTypes.func,
  /**
   * Callback for `onPaste` event
   */
  onPaste: PropTypes.func,
};

Input.defaultProps = {
  autoFocus: false,
  className: '',
  disabled: false,
  forceFocus: false,
  hint: null,
  hintClassName: '',
  id: null,
  inputClassName: '',
  label: '',
  labelClassName: '',
  placeholder: null,
  readonly: false,
  resize: false,
  rows: 2,
  status: 'normal',
  tabIndex: 0,
  type: 'text',
  value: '',
  onChange: () => {},
  onBlur: () => {},
  onFocus: () => {},
  onKeyDown: () => {},
  onKeyPress: () => {},
  onKeyUp: () => {},
  onPaste: () => {},
};

export default Input;
