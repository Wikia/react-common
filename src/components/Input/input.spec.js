import React from 'react';
import renderer from 'react-test-renderer';
import {shallow, mount} from 'enzyme';
import sinon from 'sinon';
import uniqueId from 'lodash.uniqueid';

import Input from './index';

// mock lodash's uniqueId
jest.mock('lodash.uniqueid');
uniqueId.mockImplementation(() => 'foo');

/* eslint-disable no-console */

test('Input renders correctly with default values', () => {
  const component = renderer.create(
    <Input label="an input label" />,
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('Input renders correctly disabled', () => {
  const component = renderer.create(
    <Input label="a disabled label" disabled />,
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('Input renders correctly readonly', () => {
  const component = renderer.create(
    <Input label="a readonly label" readonly />,
  );
  expect(component.toJSON()).toMatchSnapshot();
});


test('Input lifecycle works', () => {
  const component = mount(
    <Input label="foo label" value="foo" />,
  );
  expect(component).toMatchSnapshot();

  component.setProps({value: 'bar'});
  expect(component).toMatchSnapshot();

  component.setProps({id: 'bar'});
  expect(component).toMatchSnapshot();
});

test('Input renders correctly all the standard params', () => {
  const component = renderer.create(
    <Input
      id="input_label"
      label="an input label"
      value="foo"
      onChange={(value) => { console.log(value); }}
    />,
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('Input renders correctly with an error status and a hint', () => {
  const component = renderer.create(
    <Input label="an input label" value="bar" hint="an error occured" status="error" />,
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('Input renders correctly with custom class names', () => {
  const component = renderer.create(
    <Input
      label="an input with custom classes"
      className="general-class"
      hintClassName="hint-class"
      inputClassName="input-class"
      labelClassName="label-class"
    />,
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('Input renders correctly with different types', () => {
  const types = ['text', 'number', 'email', 'search', 'tel', 'url', 'password'];

  types.forEach((type) => {
    const component = renderer.create(
      <Input
        label={`an input with type ${type}`}
        type={type}
      />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});

test('Input renders correctly with tabIndex', () => {
  const component = renderer.create(
    <Input
      label="input with tabIndex"
      tabIndex={-2}
    />,
  );

  expect(component.toJSON()).toMatchSnapshot();
});

test('Input autoFocus works correctly', () => {
  const activeElementStub = sinon.stub(document, 'activeElement').get(() => null);
  const focusStub = sinon.spy(Input.prototype, 'autoFocus');

  // first `autoFocus` call on mount
  const component = mount(
    <Input
      label="autofocus"
      autoFocus
    />
  );
  expect(focusStub.callCount).toBe(1);

  // second `autoFocus` call, but `document.activeElement` will return input element
  const instance = component.instance();
  activeElementStub.get(() => instance.input);
  const inputFocusSpy = sinon.spy();
  instance.input.focus = inputFocusSpy;
  component.setProps({label: 'bar'});

  expect(inputFocusSpy.notCalled).toBe(true);
  expect(focusStub.callCount).toBe(2);

  // third `autoFocus` call, but `document.activeElement` will return null
  activeElementStub.get(() => null);
  component.setProps({label: 'foo1'});

  expect(inputFocusSpy.calledOnce).toBe(true);
  expect(focusStub.callCount).toBe(3);
});

test('Input can fire all default callbacks', () => {
  const wrapper = shallow(
    <Input label="callbacks test" />
  );

  wrapper.find('input').simulate('change', {target: {value: 'newvalue'}});
  wrapper.find('input').simulate('focus');
  wrapper.find('input').simulate('blur');
  wrapper.find('input').simulate('keyup');
  wrapper.find('input').simulate('keydown');
  wrapper.find('input').simulate('keypress');

  expect(true).toBe(true);
});

test('Input can fire all callbacks', () => {
  const mockOnBlur = sinon.spy();
  const mockOnChange = sinon.spy();
  const mockOnFocus = sinon.spy();
  const mockOnKeyDown = sinon.spy();
  const mockOnKeyPress = sinon.spy();
  const mockOnKeyUp = sinon.spy();
  const wrapper = shallow(
    <Input
      label="callbacks test (real-world)"
      onBlur={mockOnBlur}
      onChange={mockOnChange}
      onFocus={mockOnFocus}
      onKeyDown={mockOnKeyDown}
      onKeyPress={mockOnKeyPress}
      onKeyUp={mockOnKeyUp}
    />
  );

  wrapper.find('input').simulate('change', {target: {value: 'newvalue'}});
  wrapper.find('input').simulate('focus');
  wrapper.find('input').simulate('blur');
  wrapper.find('input').simulate('keyup');
  wrapper.find('input').simulate('keydown');
  wrapper.find('input').simulate('keypress');

  expect(mockOnChange.calledOnce).toBe(true);
  expect(mockOnFocus.calledOnce).toBe(true);
  expect(mockOnBlur.calledOnce).toBe(true);
  expect(mockOnKeyDown.calledOnce).toBe(true);
  expect(mockOnKeyPress.calledOnce).toBe(true);
  expect(mockOnKeyUp.calledOnce).toBe(true);
});


test('Input does not call onChange and onFocus callbacks on readonly', () => {
  const mockOnChange = sinon.spy();
  const mockOnFocus = sinon.spy();
  const wrapper = shallow(
    <Input
      label="change test for readonly"
      onChange={mockOnChange}
      onFocus={mockOnFocus}
      readonly
    />
  );

  wrapper.find('input').simulate('change', {target: {value: 'newvalue'}});
  wrapper.find('input').simulate('focus');

  expect(mockOnChange.calledOnce).toBe(false);
  expect(mockOnFocus.calledOnce).toBe(false);
});

test('Input can call focus and blur', () => {
  const activeElementStub = sinon.stub(document, 'activeElement').get(() => null);

  const focusStub = sinon.spy();
  const blurStub = sinon.spy();

  const wrapper = mount(
    <Input
      label="focus and blur"
    />
  );

  const instance = wrapper.instance();

  instance.input.focus = focusStub;
  instance.input.blur = blurStub;

  // correct - focus and blur
  instance.focus();
  instance.blur();

  // already focused, can't focus again
  activeElementStub.get(() => instance.input);
  instance.focus();

  // incorrect - lost input
  instance.input = null;
  instance.focus();
  instance.blur();

  expect(focusStub.calledOnce).toBe(true);
  expect(blurStub.calledOnce).toBe(true);
});
