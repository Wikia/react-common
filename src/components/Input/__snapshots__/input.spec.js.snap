// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input autoresizing works correctly 1`] = `
<Input
  autoFocus={false}
  className=""
  disabled={false}
  forceFocus={false}
  hint={null}
  hintClassName=""
  id={null}
  inputClassName=""
  label="Autoresize"
  labelClassName=""
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyPress={[Function]}
  onKeyUp={[Function]}
  onPaste={[Function]}
  placeholder={null}
  readonly={false}
  resize="auto"
  rows={2}
  status="normal"
  tabIndex={0}
  type="multiline"
  value="This is some value"
>
  <div
    className="wds-input"
  >
    <textarea
      className="wds-input__field "
      disabled={false}
      id="foo"
      name="foo"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onInput={[Function]}
      onKeyDown={[Function]}
      onKeyPress={[Function]}
      onKeyUp={[Function]}
      onPaste={[Function]}
      placeholder={null}
      readOnly={false}
      rows={2}
      tabIndex={0}
      value="This is some value"
    >
      This is some value
    </textarea>
    <label
      className="wds-input__label "
      htmlFor="foo"
    >
      Autoresize
    </label>
  </div>
</Input>
`;

exports[`Input autoresizing works correctly 2`] = `
<Input
  autoFocus={false}
  className=""
  disabled={false}
  forceFocus={false}
  hint={null}
  hintClassName=""
  id={null}
  inputClassName=""
  label="Autoresize"
  labelClassName=""
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyPress={[Function]}
  onKeyUp={[Function]}
  onPaste={[Function]}
  placeholder={null}
  readonly={false}
  resize="auto"
  rows={2}
  status="normal"
  tabIndex={0}
  type="multiline"
  value="This is some value"
>
  <div
    className="wds-input"
  >
    <textarea
      className="wds-input__field "
      disabled={false}
      id="foo"
      name="foo"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onInput={[Function]}
      onKeyDown={[Function]}
      onKeyPress={[Function]}
      onKeyUp={[Function]}
      onPaste={[Function]}
      placeholder={null}
      readOnly={false}
      rows={2}
      style={
        Object {
          "height": "101px",
        }
      }
      tabIndex={0}
      value="This is some value"
    >
      This is some value
    </textarea>
    <label
      className="wds-input__label "
      htmlFor="foo"
    >
      Autoresize
    </label>
  </div>
</Input>
`;

exports[`Input lifecycle works 1`] = `
<Input
  autoFocus={false}
  className=""
  disabled={false}
  forceFocus={false}
  hint={null}
  hintClassName=""
  id={null}
  inputClassName=""
  label="foo label"
  labelClassName=""
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyPress={[Function]}
  onKeyUp={[Function]}
  onPaste={[Function]}
  placeholder={null}
  readonly={false}
  resize={false}
  rows={2}
  status="normal"
  tabIndex={0}
  type="text"
  value="foo"
>
  <div
    className="wds-input"
  >
    <input
      className="wds-input__field "
      disabled={false}
      id="foo"
      name="foo"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyPress={[Function]}
      onKeyUp={[Function]}
      onPaste={[Function]}
      placeholder={null}
      readOnly={false}
      tabIndex={0}
      type="text"
      value="foo"
    />
    <label
      className="wds-input__label "
      htmlFor="foo"
    >
      foo label
    </label>
  </div>
</Input>
`;

exports[`Input lifecycle works 2`] = `
<Input
  autoFocus={false}
  className=""
  disabled={false}
  forceFocus={false}
  hint={null}
  hintClassName=""
  id={null}
  inputClassName=""
  label="foo label"
  labelClassName=""
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyPress={[Function]}
  onKeyUp={[Function]}
  onPaste={[Function]}
  placeholder={null}
  readonly={false}
  resize={false}
  rows={2}
  status="normal"
  tabIndex={0}
  type="text"
  value="bar"
>
  <div
    className="wds-input"
  >
    <input
      className="wds-input__field "
      disabled={false}
      id="foo"
      name="foo"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyPress={[Function]}
      onKeyUp={[Function]}
      onPaste={[Function]}
      placeholder={null}
      readOnly={false}
      tabIndex={0}
      type="text"
      value="bar"
    />
    <label
      className="wds-input__label "
      htmlFor="foo"
    >
      foo label
    </label>
  </div>
</Input>
`;

exports[`Input lifecycle works 3`] = `
<Input
  autoFocus={false}
  className=""
  disabled={false}
  forceFocus={false}
  hint={null}
  hintClassName=""
  id="bar"
  inputClassName=""
  label="foo label"
  labelClassName=""
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyPress={[Function]}
  onKeyUp={[Function]}
  onPaste={[Function]}
  placeholder={null}
  readonly={false}
  resize={false}
  rows={2}
  status="normal"
  tabIndex={0}
  type="text"
  value="bar"
>
  <div
    className="wds-input"
  >
    <input
      className="wds-input__field "
      disabled={false}
      id="bar"
      name="bar"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyPress={[Function]}
      onKeyUp={[Function]}
      onPaste={[Function]}
      placeholder={null}
      readOnly={false}
      tabIndex={0}
      type="text"
      value="bar"
    />
    <label
      className="wds-input__label "
      htmlFor="bar"
    >
      foo label
    </label>
  </div>
</Input>
`;

exports[`Input renders correctly all the standard params 1`] = `
<div
  className="wds-input"
>
  <input
    className="wds-input__field "
    disabled={false}
    id="input_label"
    name="input_label"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    onKeyUp={[Function]}
    onPaste={[Function]}
    placeholder={null}
    readOnly={false}
    tabIndex={0}
    type="text"
    value="foo"
  />
  <label
    className="wds-input__label "
    htmlFor="input_label"
  >
    an input label
  </label>
</div>
`;

exports[`Input renders correctly disabled 1`] = `
<div
  className="wds-input is-disabled is-empty"
>
  <input
    className="wds-input__field "
    disabled={true}
    id="foo"
    name="foo"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    onKeyUp={[Function]}
    onPaste={[Function]}
    placeholder={null}
    readOnly={false}
    tabIndex={0}
    type="text"
    value=""
  />
  <label
    className="wds-input__label "
    htmlFor="foo"
  >
    a disabled label
  </label>
</div>
`;

exports[`Input renders correctly readonly 1`] = `
<div
  className="wds-input is-readonly is-empty"
>
  <input
    className="wds-input__field "
    disabled={false}
    id="foo"
    name="foo"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    onKeyUp={[Function]}
    onPaste={[Function]}
    placeholder={null}
    readOnly={true}
    tabIndex={0}
    type="text"
    value=""
  />
  <label
    className="wds-input__label "
    htmlFor="foo"
  >
    a readonly label
  </label>
</div>
`;

exports[`Input renders correctly with a placeholder 1`] = `
<div
  className="wds-input is-empty"
>
  <input
    className="wds-input__field "
    disabled={false}
    id="foo"
    name="foo"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    onKeyUp={[Function]}
    onPaste={[Function]}
    placeholder="That's a placeholder"
    readOnly={false}
    tabIndex={0}
    type="text"
    value=""
  />
  <label
    className="wds-input__label "
    htmlFor="foo"
  >
    
  </label>
</div>
`;

exports[`Input renders correctly with an error status and a hint 1`] = `
<div
  className="wds-input has-error has-hint"
>
  <input
    className="wds-input__field "
    disabled={false}
    id="foo"
    name="foo"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    onKeyUp={[Function]}
    onPaste={[Function]}
    placeholder={null}
    readOnly={false}
    tabIndex={0}
    type="text"
    value="bar"
  />
  <div
    className="wds-input__hint "
  >
    an error occured
  </div>
  <label
    className="wds-input__label "
    htmlFor="foo"
  >
    an input label
  </label>
</div>
`;

exports[`Input renders correctly with custom class names 1`] = `
<div
  className="wds-input is-empty general-class"
>
  <input
    className="wds-input__field input-class"
    disabled={false}
    id="foo"
    name="foo"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    onKeyUp={[Function]}
    onPaste={[Function]}
    placeholder={null}
    readOnly={false}
    tabIndex={0}
    type="text"
    value=""
  />
  <label
    className="wds-input__label label-class"
    htmlFor="foo"
  >
    an input with custom classes
  </label>
</div>
`;

exports[`Input renders correctly with default values 1`] = `
<div
  className="wds-input is-empty"
>
  <input
    className="wds-input__field "
    disabled={false}
    id="foo"
    name="foo"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    onKeyUp={[Function]}
    onPaste={[Function]}
    placeholder={null}
    readOnly={false}
    tabIndex={0}
    type="text"
    value=""
  />
  <label
    className="wds-input__label "
    htmlFor="foo"
  >
    an input label
  </label>
</div>
`;

exports[`Input renders correctly with different types 1`] = `
<div
  className="wds-input"
>
  <input
    className="wds-input__field "
    disabled={false}
    id="foo"
    name="foo"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    onKeyUp={[Function]}
    onPaste={[Function]}
    placeholder={null}
    readOnly={false}
    tabIndex={0}
    type="text"
    value="an input with value text"
  />
  <label
    className="wds-input__label "
    htmlFor="foo"
  >
    an input with type text
  </label>
</div>
`;

exports[`Input renders correctly with different types 2`] = `
<div
  className="wds-input"
>
  <input
    className="wds-input__field "
    disabled={false}
    id="foo"
    name="foo"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    onKeyUp={[Function]}
    onPaste={[Function]}
    placeholder={null}
    readOnly={false}
    tabIndex={0}
    type="number"
    value="an input with value number"
  />
  <label
    className="wds-input__label "
    htmlFor="foo"
  >
    an input with type number
  </label>
</div>
`;

exports[`Input renders correctly with different types 3`] = `
<div
  className="wds-input"
>
  <input
    className="wds-input__field "
    disabled={false}
    id="foo"
    name="foo"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    onKeyUp={[Function]}
    onPaste={[Function]}
    placeholder={null}
    readOnly={false}
    tabIndex={0}
    type="email"
    value="an input with value email"
  />
  <label
    className="wds-input__label "
    htmlFor="foo"
  >
    an input with type email
  </label>
</div>
`;

exports[`Input renders correctly with different types 4`] = `
<div
  className="wds-input"
>
  <input
    className="wds-input__field "
    disabled={false}
    id="foo"
    name="foo"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    onKeyUp={[Function]}
    onPaste={[Function]}
    placeholder={null}
    readOnly={false}
    tabIndex={0}
    type="search"
    value="an input with value search"
  />
  <label
    className="wds-input__label "
    htmlFor="foo"
  >
    an input with type search
  </label>
</div>
`;

exports[`Input renders correctly with different types 5`] = `
<div
  className="wds-input"
>
  <input
    className="wds-input__field "
    disabled={false}
    id="foo"
    name="foo"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    onKeyUp={[Function]}
    onPaste={[Function]}
    placeholder={null}
    readOnly={false}
    tabIndex={0}
    type="tel"
    value="an input with value tel"
  />
  <label
    className="wds-input__label "
    htmlFor="foo"
  >
    an input with type tel
  </label>
</div>
`;

exports[`Input renders correctly with different types 6`] = `
<div
  className="wds-input"
>
  <input
    className="wds-input__field "
    disabled={false}
    id="foo"
    name="foo"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    onKeyUp={[Function]}
    onPaste={[Function]}
    placeholder={null}
    readOnly={false}
    tabIndex={0}
    type="url"
    value="an input with value url"
  />
  <label
    className="wds-input__label "
    htmlFor="foo"
  >
    an input with type url
  </label>
</div>
`;

exports[`Input renders correctly with different types 7`] = `
<div
  className="wds-input"
>
  <input
    className="wds-input__field "
    disabled={false}
    id="foo"
    name="foo"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    onKeyUp={[Function]}
    onPaste={[Function]}
    placeholder={null}
    readOnly={false}
    tabIndex={0}
    type="password"
    value="an input with value password"
  />
  <label
    className="wds-input__label "
    htmlFor="foo"
  >
    an input with type password
  </label>
</div>
`;

exports[`Input renders correctly with multiline 1`] = `
<div
  className="wds-input"
>
  <textarea
    className="wds-input__field "
    disabled={false}
    id="foo"
    name="foo"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    onKeyUp={[Function]}
    onPaste={[Function]}
    placeholder={null}
    readOnly={false}
    rows={3}
    tabIndex={0}
    value="this is multiline"
  >
    this is multiline
  </textarea>
  <label
    className="wds-input__label "
    htmlFor="foo"
  >
    Textarea
  </label>
</div>
`;

exports[`Input renders correctly with multiline 2`] = `
<div
  className="wds-input is-resize"
>
  <textarea
    className="wds-input__field "
    disabled={false}
    id="foo"
    name="foo"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    onKeyUp={[Function]}
    onPaste={[Function]}
    placeholder={null}
    readOnly={false}
    rows={3}
    tabIndex={0}
    value="this is multiline (with resize)"
  >
    this is multiline (with resize)
  </textarea>
  <label
    className="wds-input__label "
    htmlFor="foo"
  >
    Textarea
  </label>
</div>
`;

exports[`Input renders correctly with tabIndex 1`] = `
<div
  className="wds-input is-empty"
>
  <input
    className="wds-input__field "
    disabled={false}
    id="foo"
    name="foo"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    onKeyUp={[Function]}
    onPaste={[Function]}
    placeholder={null}
    readOnly={false}
    tabIndex={-2}
    type="text"
    value=""
  />
  <label
    className="wds-input__label "
    htmlFor="foo"
  >
    input with tabIndex
  </label>
</div>
`;
